<a href="https://colab.research.google.com/github/Rossel/DataQuest_Courses/blob/master/060__Group_Summary_Statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# COURSE 1/5: SQL FUNDAMENTALS

# MISSION 3: Group Summary Statistics

*Learn how to compute statistics across groups*

In this mission, we cover the CASE, GROUP BY and HAVING clauses. We can use these to quickly calculate powerful summary statistics in SQL and program if/then logic. 

## 1. Introduction

In the last mission, we computed summary statistics across columns with SQL. In many cases, though, we want to drill down even more and compute summary statistics per group to answer questions like:

- What is the total number of graduates by major category?
- What major category pays better?
- How does the sample size vary by major category?

In this mission, we'll answer questions like these by learning how to calculate summary statistics by groups.

We'll continue working with the `recent_grads` table of `jobs.db`. Recall that each row represents a single college major, and contains information about post-graduation employment of students who studied the major. Here are some descriptions for just a few of the 21 total columns:

- `Rank` - The major's rank by median earnings
- `Major` - The name of the major
- `Major_category` - The broader category the major belongs to
- `Total` - The total number of people who graduated the major
- `Sample_size` - The size of the sample
- `Men` - The number of male graduates
- `Women` - The number of female graduates
- `ShareWomen` - Women as a proportion of the total number of graduates (a number ranging from `0` to `1`)
- `Employed` - The number of employed graduates

Let's load the database file and display the first few rows and columns in the dataset:



In [1]:
# Import functions from Google modules into Colaboratory
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [2]:
# Insert file id from Google Drive shareable link:
# https://drive.google.com/file/d/1dGtg_Gx7VZgH4ZVydd2sFly3tNLsvhds/view?usp=sharing
id = '1dGtg_Gx7VZgH4ZVydd2sFly3tNLsvhds'

In [3]:
# Download the dataset
downloaded = drive.CreateFile({'id':id}) 
downloaded.GetContentFile('jobs.db')

In [4]:
# Import SQLite3 and pandas library
import sqlite3
import pandas as pd

In [5]:
%%capture
%load_ext sql
%sql sqlite:///jobs.db

In [6]:
%%sql

  SELECT * 
  FROM recent_grads
  LIMIT 5

 * sqlite:///jobs.db
Done.


[(0, 1, 2419, 'PETROLEUM ENGINEERING', 'Engineering', 2339, 36, 2057, 282, 0.120564344, 1976, 1849, 270, 1207, 37, 0.018380527, 110000, 95000, 125000, 1534, 364, 193),
 (1, 2, 2416, 'MINING AND MINERAL ENGINEERING', 'Engineering', 756, 7, 679, 77, 0.10185185199999999, 640, 556, 170, 388, 85, 0.117241379, 75000, 55000, 90000, 350, 257, 50),
 (2, 3, 2415, 'METALLURGICAL ENGINEERING', 'Engineering', 856, 3, 725, 131, 0.153037383, 648, 558, 133, 340, 16, 0.024096386, 73000, 50000, 105000, 456, 176, 0),
 (3, 4, 2417, 'NAVAL ARCHITECTURE AND MARINE ENGINEERING', 'Engineering', 1258, 16, 1123, 135, 0.107313196, 758, 1069, 150, 692, 40, 0.050125313, 70000, 43000, 80000, 529, 102, 0),
 (4, 5, 2405, 'CHEMICAL ENGINEERING', 'Engineering', 32260, 289, 21239, 11021, 0.341630502, 25694, 23170, 5180, 16697, 1672, 0.061097712, 65000, 50000, 75000, 18314, 4440, 972)]

Before we dive into answer questions like the above, we'll learn how to code [if/then logic](https://en.wikipedia.org/wiki/Conditional_(computer_programming)) in SQL.

## 2. If/Then in SQL

It is very common to want to create new columns according to rules based on other columns. For example, in order to quickly identify what are the top 10 majors (as per the `rank` column), we may want to create a column that indicates whether the `value` on rank for each row is greater than 10 or not.

![alt text](https://dq-content.s3.amazonaws.com/254/2.1-rank.gif)



We can do such things with the `CASE` clause in SQL. We'll program the following logic:

- **When** the row corresponds to a top 10 major, **then** assign the value `Top 10`.
- **When** the row corresponds to a top 20 major that isn't a top 10 major, **then** assign the value `Top 20`.
- **Otherwise** let it be a missing value.

We can do this with the following syntax:
```
SELECT CASE
       WHEN rank <= 10 THEN 'Top 10'
       WHEN rank <= 20 THEN 'Top 20'
       ELSE NULL
       END AS rank_category
  FROM recent_grads;
```

A more general syntax looks like this:
```
CASE
WHEN <condition_1> THEN <value_1>
WHEN <condition_2> THEN <value_2>
ELSE <value_3>
END AS <new_column_name>
```

Let's look at what the results look like. We'll look at a few select rows of the result of the following query:



In [7]:
%%sql

SELECT Major, Rank,
       CASE
       WHEN rank <= 10 THEN 'Top 10'
       WHEN rank <= 20 THEN 'Top 20'
       ELSE NULL
       END AS rank_category
  FROM recent_grads
 LIMIT 25;

 * sqlite:///jobs.db
Done.


[('PETROLEUM ENGINEERING', 1, 'Top 10'),
 ('MINING AND MINERAL ENGINEERING', 2, 'Top 10'),
 ('METALLURGICAL ENGINEERING', 3, 'Top 10'),
 ('NAVAL ARCHITECTURE AND MARINE ENGINEERING', 4, 'Top 10'),
 ('CHEMICAL ENGINEERING', 5, 'Top 10'),
 ('NUCLEAR ENGINEERING', 6, 'Top 10'),
 ('ACTUARIAL SCIENCE', 7, 'Top 10'),
 ('ASTRONOMY AND ASTROPHYSICS', 8, 'Top 10'),
 ('MECHANICAL ENGINEERING', 9, 'Top 10'),
 ('ELECTRICAL ENGINEERING', 10, 'Top 10'),
 ('COMPUTER ENGINEERING', 11, 'Top 20'),
 ('AEROSPACE ENGINEERING', 12, 'Top 20'),
 ('BIOMEDICAL ENGINEERING', 13, 'Top 20'),
 ('MATERIALS SCIENCE', 14, 'Top 20'),
 ('ENGINEERING MECHANICS PHYSICS AND SCIENCE', 15, 'Top 20'),
 ('BIOLOGICAL ENGINEERING', 16, 'Top 20'),
 ('INDUSTRIAL AND MANUFACTURING ENGINEERING', 17, 'Top 20'),
 ('GENERAL ENGINEERING', 18, 'Top 20'),
 ('ARCHITECTURAL ENGINEERING', 19, 'Top 20'),
 ('COURT REPORTING', 20, 'Top 20'),
 ('COMPUTER SCIENCE', 21, None),
 ('FOOD SCIENCE', 22, None),
 ('ELECTRICAL ENGINEERING TECHNOLOGY', 23, None),
 ('MATERIALS ENGINEERING AND MATERIALS SCIENCE', 24, None),
 ('MANAGEMENT INFORMATION SYSTEMS AND STATISTICS', 25, None)]

We can see that our logic was coded correctly. We'll dissect this clause in the next screen.

In this exercise we'll categorize the `Sample_size` column into three categories: `Small`, `Medium`, and `Large`.

The size of a sample is of extreme importance for statistical analysis, you can learn more about this in the statistics courses.





**Instructions:**

Write a SQL query that displays, with the alias `Sample_category`, the column whose values are given by the following rules:

- `Small` if `Sample_size` is smaller than 200
- `Medium` if `Sample_size` is equal to or higher than 200, and smaller than 1000
- `Large` if `Sample_size` is equal to or higher than 1000


In [20]:
%%sql

SELECT CASE
  WHEN Sample_size < 200 THEN 'Small'
  WHEN Sample_size < 1000 THEN 'Medium'
  ELSE 'Large'
   END AS Sample_category
  FROM recent_grads
 LIMIT 10;

 * sqlite:///jobs.db
Done.


[('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Large',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Large',),
 ('Large',),
 ('Large',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Large',),
 ('Large',),
 ('Large',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Large',),
 ('Large',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Large',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Large',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Large',),
 ('Large',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Large',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Medium',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',),
 ('Small',)]

## 3. Dissecting CASE

Let's recall the syntax we saw in the previous screen.

```
CASE
WHEN <condition_1> THEN <value_1>
WHEN <condition_2> THEN <value_2>
ELSE <value_3>
END AS <new_column_name>
```

Note the following:

- It starts with `CASE` to indicate that we'll be creating values by cases.
- It ends with `END` to indicate where the `CASE` clause terminates.
- Each explicit case is signaled by the reserved word `WHEN`.
- The value for each case is given after the reserved word `THEN`.
- There is a fallback value given by the reserved word `ELSE`.


Here are some very important observations:

- Whatever you can use in `WHERE` for filtering, you can also use in place of the conditions above.
- There can be one or more `WHEN` lines. We exemplified with three, but it works just with one, two or, for instance, ten.
- The `ELSE` line is optional — without it, rows that don't match any `WHEN` will be set with a missing value (`NULL`).

**Instructions:**

Write a SQL query that displays in order `Major`, `Sample_size` and a column named `Sample_category`, whose values are defined by the following rules:

- `Small` if `Sample_size` is smaller than 200
- `Medium` if `Sample_size` is equal to or higher than 200, and smaller than 1000
- `Large` if `Sample_size` is equal to or higher than 1000

In [9]:
%%sql

SELECT Major, Sample_size,
       CASE
       WHEN Sample_size < 200 THEN 'Small'
       WHEN Sample_size < 1000 THEN 'Medium'
       ELSE 'Large'
       END AS Sample_category
  FROM recent_grads
 LIMIT 15;

 * sqlite:///jobs.db
Done.


[('PETROLEUM ENGINEERING', 36, 'Small'),
 ('MINING AND MINERAL ENGINEERING', 7, 'Small'),
 ('METALLURGICAL ENGINEERING', 3, 'Small'),
 ('NAVAL ARCHITECTURE AND MARINE ENGINEERING', 16, 'Small'),
 ('CHEMICAL ENGINEERING', 289, 'Medium'),
 ('NUCLEAR ENGINEERING', 17, 'Small'),
 ('ACTUARIAL SCIENCE', 51, 'Small'),
 ('ASTRONOMY AND ASTROPHYSICS', 10, 'Small'),
 ('MECHANICAL ENGINEERING', 1029, 'Large'),
 ('ELECTRICAL ENGINEERING', 631, 'Medium'),
 ('COMPUTER ENGINEERING', 399, 'Medium'),
 ('AEROSPACE ENGINEERING', 147, 'Small'),
 ('BIOMEDICAL ENGINEERING', 79, 'Small'),
 ('MATERIALS SCIENCE', 22, 'Small'),
 ('ENGINEERING MECHANICS PHYSICS AND SCIENCE', 30, 'Small')]

## 4. Calculating Group-Level Summary Statistics

Let's go back to calculating summary statistics by group. By group we mean any of the unique values in a column. Typically group statistics are calculated only for columns in which the values represent categories (as opposed to measurements of some sort — like age, currency, and so on).

The `GROUP BY` SQL statement allows us to compute summary statistics by group. When we use this statement, SQL creates a group for each unique value in a column or set of columns (the same values we get when we use the `DISTINCT` statement), and then does the calculations for them).

To illustrate, we can find the total number of people employed in each major category with the following query below. This will give us the total number of employed graduates for each major category. Here's a truncated view of the output:

In [10]:
%%sql

SELECT SUM(Employed) 
  FROM recent_grads 
 GROUP BY Major_category
 LIMIT 3;

 * sqlite:///jobs.db
Done.


[(66943,), (288114,), (302797,)]

The output shows aggregate counts of the `Employed` column for each `Major_category`. Unfortunately, it doesn't indicate which major category each row refers to. We can fix this by including the `Major_category` column in our query:

In [11]:
%%sql

SELECT Major_category, SUM(Employed) 
  FROM recent_grads 
 GROUP BY Major_category;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 66943),
 ('Arts', 288114),
 ('Biology & Life Science', 302797),
 ('Business', 1088742),
 ('Communications & Journalism', 330660),
 ('Computers & Mathematics', 237894),
 ('Education', 479839),
 ('Engineering', 420372),
 ('Health', 372147),
 ('Humanities & Liberal Arts', 544118),
 ('Industrial Arts & Consumer Services', 189043),
 ('Interdisciplinary', 9821),
 ('Law & Public Policy', 144790),
 ('Physical Sciences', 139231),
 ('Psychology & Social Work', 380344),
 ('Social Science', 401493)]

One of the questions we put forth at beginning of the mission was: What is the total number of graduates by major category? Let's answer it.

**Instructions:**

Write a SQL query that, for each major category, displays:

- The `major` category
- The total number of graduates for `Major_category` with the alias `Total_graduates`

In [12]:
%%sql

SELECT Major_category, SUM(Total) AS Total_graduates
  FROM recent_grads
 GROUP BY Major_category;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 79981),
 ('Arts', 357130),
 ('Biology & Life Science', 453862),
 ('Business', 1302376),
 ('Communications & Journalism', 392601),
 ('Computers & Mathematics', 299008),
 ('Education', 559129),
 ('Engineering', 537583),
 ('Health', 463230),
 ('Humanities & Liberal Arts', 713468),
 ('Industrial Arts & Consumer Services', 229792),
 ('Interdisciplinary', 12296),
 ('Law & Public Policy', 179107),
 ('Physical Sciences', 185479),
 ('Psychology & Social Work', 481007),
 ('Social Science', 529966)]

## 5. GROUP BY Visual Breakdown

Let's recall the query from the example in the last screen.

In [13]:
%%sql

SELECT Major_category, SUM(Employed) 
  FROM recent_grads 
 GROUP BY Major_category
 LIMIT 3;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 66943),
 ('Arts', 288114),
 ('Biology & Life Science', 302797)]

Here's how the query works. The `GROUP BY` statement splits the `Major_category` column into groups (with one group for each unique major category), then calculates the sum for each group. The following diagram shows how `GROUP BY` splits the data (the diagram uses a small sample from the `recent_grads` table):

![alt text](https://dq-content.s3.amazonaws.com/254/5.1-group.gif)

For each group, the `GROUP BY` statement queries each column, and runs all of the aggregation functions we include in the query after the `SELECT` statement:

![alt text](https://dq-content.s3.amazonaws.com/254/5.2-sum.gif)

If a column is selected, the SQL engine will use the last value for that column in the group. If an aggregation function is selected, the SQL engine will compute the value for that aggregation function across the group.

The query in the diagram will give us the following result:


Employed|	Major_category	|SUM(Employed)
---| ---|---
1290	|Agriculture|	4439
36165	|Arts|	39075

Now let's try using a different aggregate function.

**Instructions:**

Write a SQL query that, for each major category, displays:
- The major category
- The average share of women with the alias `Average_women`.

In [14]:
%%sql

SELECT Major_category,
       AVG(ShareWomen) AS Average_women
  FROM recent_grads
 GROUP BY Major_category;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 0.6179384232),
 ('Arts', 0.56185119575),
 ('Biology & Life Science', 0.584518475857143),
 ('Business', 0.4050631853076923),
 ('Communications & Journalism', 0.64383484025),
 ('Computers & Mathematics', 0.5127519954545455),
 ('Education', 0.6749855163125),
 ('Engineering', 0.2571578951034483),
 ('Health', 0.6168565694166667),
 ('Humanities & Liberal Arts', 0.6761934042),
 ('Industrial Arts & Consumer Services', 0.4493512688571429),
 ('Interdisciplinary', 0.495397153),
 ('Law & Public Policy', 0.3359896912),
 ('Physical Sciences', 0.5087494197),
 ('Psychology & Social Work', 0.7777631628888888),
 ('Social Science', 0.5390672957777778)]

## 6. Multiple Summary Statistics by Group

Just like we were able to compute multiple summary statistics for the whole table in the last mission, we can also compute multiple summary statistics by groups.

Continuing from our example, here's a query that for each major category finds:

- The total number of employed graduates
- The average number of employed graduates
- The maximum number of employed graduates in a major
- The minimum number of employed graduates in a major

In [15]:
%%sql

SELECT Major_category,
       SUM(Employed), AVG(Employed), MAX(Employed), MIN(Employed)
  FROM recent_grads
 GROUP BY Major_category
 LIMIT 3;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 66943, 6694.3, 17112, 613),
 ('Arts', 288114, 36014.25, 83483, 2914),
 ('Biology & Life Science', 302797, 21628.35714285714, 182295, 1010)]

We can also make calculations with the resulting columns, much like you learned before. Here's a query that computes the difference between the maximum and the minimum number of employed graduates:

In [16]:
%%sql

SELECT Major_category,
       SUM(Employed), AVG(Employed), MAX(Employed) - MIN(Employed) as range_employed
  FROM recent_grads
 GROUP BY Major_category;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 66943, 6694.3, 16499),
 ('Arts', 288114, 36014.25, 80569),
 ('Biology & Life Science', 302797, 21628.35714285714, 181285),
 ('Business', 1088742, 83749.38461538461, 273322),
 ('Communications & Journalism', 330660, 82665.0, 134954),
 ('Computers & Mathematics', 237894, 21626.727272727272, 101528),
 ('Education', 479839, 29989.9375, 148636),
 ('Engineering', 420372, 14495.586206896553, 75838),
 ('Health', 372147, 31012.25, 173851),
 ('Humanities & Liberal Arts', 544118, 36274.53333333333, 146393),
 ('Industrial Arts & Consumer Services', 189043, 27006.14285714286, 103078),
 ('Interdisciplinary', 9821, 9821.0, 0),
 ('Law & Public Policy', 144790, 28958.0, 124463),
 ('Physical Sciences', 139231, 13923.1, 47389),
 ('Psychology & Social Work', 380344, 42260.444444444445, 306853),
 ('Social Science', 401493, 44610.333333333336, 130727)]

Note that we were able to use an alias for the result of the calculation without an issue. Aliases are used in the same way that you learned before.

Now that we have a better understanding of the `GROUP BY` statement, let's practice using it by computing summary statistics by group for the `recent_grads` table.

**Instructions:**

Write a query that, for each major category, displays in order:
- The major category
- The total amount of women with the alias `Total_women`
- The average percentage of women with the alias `Mean_women`
- The result of multiplying the total number of graduates by the average percentage of women with the alias `Estimate_women`.

In [17]:
%%sql

SELECT Major_category, 
       SUM(Women) as 'Total_women',
       AVG(ShareWomen) as 'Mean_women',
       SUM(Total) * AVG(ShareWomen) as 'Estimate_women'
  FROM recent_grads
 GROUP BY Major_category;

 * sqlite:///jobs.db
Done.


[('Agriculture & Natural Resources', 249812, 0.6179384232, 49423.333025959204),
 ('Arts', 140469, 0.56185119575, 200653.91753819748),
 ('Biology & Life Science', 578132, 0.584518475857143, 265290.7244894746),
 ('Business', 110367, 0.4050631853076923, 527544.5710282911),
 ('Communications & Journalism', 98278, 0.64383484025, 252770.20211699026),
 ('Computers & Mathematics', 62599, 0.5127519954545455, 153316.94865687276),
 ('Education', 612958, 0.6749855163125, 377403.9767502918),
 ('Engineering', 118051, 0.2571578951034483, 138243.71272339704),
 ('Health', 312026, 0.6168565694166667, 285746.4686508825),
 ('Humanities & Liberal Arts', 349636, 0.6761934042, 482442.35570776556),
 ('Industrial Arts & Consumer Services', 36824, 0.4493512688571429, 103257.32677322057),
 ('Interdisciplinary', 9848, 0.495397153, 6091.403393288),
 ('Law & Public Policy', 5478, 0.3359896912, 60178.1056217584),
 ('Physical Sciences', 244996, 0.5087494197, 94362.3336165363),
 ('Psychology & Social Work', 504662, 0.7777631628888888, 374109.52569169574),
 ('Social Science', 463616, 0.5390672957777778, 285687.33847416576)]

## 7. Multiple Group Columns

So far we've only grouped by one column (`Major_category`), but it's also possible to group by multiple columns.

We'll exemplify with the following table called `fruit`. This table is not in the database that we're using here, so you won't be able to experiment with it.


fruit|	color|	sourness|	weight|	price|
---|---|---|---|---
Orange|	Orange|	Sour|	131|	4
Banana|	Yellow|	Sweet|	150|	4
Papaya|	Orange|	Sweet|	450|	2
Lime|	Green|	Sour|	44|	1
Strawberry|	Red|	Sweet|	12|	3
Grapefruit|	Orange|	Sour|	236|	1
Cherry|	Red|	Sour|	12|	3
Red apple|	Red|	Sweet|	180|	5
Green apple|	Green|	Sour|	170|	2

We see that we have the groups in `color` and in `sourness`.

We've ran queries like the one below, which gives the mean weight and price.

```
SELECT color,
       AVG(weight),
       AVG(price)
  FROM fruit
 GROUP BY color;
```


color	|AVG(weight)|	AVG(price)
---|---|---
Green	|107.0|	1.5
Orange	|272.333333|	2.333333
Red	|68.0|	3.666667
Yellow|	150.0|	4.0

What if we want the mean weight and price for the group of sweet red fruits? Or more generally, the mean weight and price for all the combinations of groups of color and sourness? We can answer such questions with GROUP BY as well. Here's an example:

```
SELECT color, sourness
       AVG(weight), AVG(price), MIN(weight), MIN(price)
  FROM fruit
 GROUP BY color, sourness;
```


color|	sourness	|AVG(weight)	|AVG(price)|	MIN(weight)|	MIN(price)
---|---|---|---|---|---|
Green|	Sour|	107.0|	1.5|	44|	1
Orange|	Sour|	183.5|	2.5|	131	|1
Orange|	Sweet|	450.0|	2.0|	450|	2
Red|	Sour|	12.0|	3.0|	12|	3
Red|	Sweet|	96.0	|4.0|	12|	3
Yellow|	Sweet|	150.0|	4.0|	150|	4


We can easily confirm the minimum values by spotchecking.

Earlier in the mission we created a column whose values are given by the following rules:

- `Small` if `Sample_size` is smaller than 200
- `Medium` if `Sample_size` is equal to or higher than 200, and smaller than 1000
- `Large` if `Sample_size` is equal to or higher than 1000

We duplicated the `recent_grads` table and added that column to a new table. It is called `new_grads` and we'll use it for the remainder of this mission.

**Instructions:**

Write a query that, for each combination of `Major_category` and `Sample_category`, displays in order:

- The major category
- The sample category
- The average percentage of women with the alias `Mean_women`
- The sum total of graduates with the alias `Total_graduates`


```
SELECT Major_category, Sample_category,
       AVG(ShareWomen) AS Mean_women,
       SUM(Total) AS Total_graduates
  FROM new_grads
 GROUP BY Major_category, Sample_category;
 ```

 [42 rows x 4 columns]

Major_category| Sample_category|Mean_women|Total_graduates
---|---|---|---|
Agriculture & Natural Resources|Medium|0.75257011|35813
Agriculture & Natural Resources|Small|0.5842805015|44168
Arts|Large|0.37435562299999997|103480

## 8. Querying Virtual Columns With the HAVING Statement

Sometimes we want to select a subset of rows after performing a `GROUP BY` query. On the last screen, for instance, we may have wanted to select only those rows where `Mean_women` is greater than `.8`.

We can't use the `WHERE` clause to do this:
```
SELECT Major_category, Sample_category,
       AVG(ShareWomen) AS Mean_women,
       SUM(Total) AS Total_graduates
  FROM new_grads
 GROUP BY Major_category, Sample_category
 WHERE Mean_women > 0.8;
 ```

 ```

(sqlite3.OperationalError) near "WHERE": syntax error
[SQL: SELECT Major_category, Sample_category,        AVG(ShareWomen) AS Mean_women,        SUM(Total) AS Total_graduates   FROM new_grads  GROUP BY Major_category, Sample_category  WHERE Mean_women > 0.8  ;]
(Background on this error at: http://sqlalche.me/e/e3q8)

```

We got an error! We'll understand why `WHERE` doesn't work here in the next chapter, for now let's learn how we would be able to subset the data.

**When we want to filter on a column generated by a `GROUP BY` query, we can use the `HAVING` statement. Here's an example:**

```
SELECT Major_category, Sample_category,
       AVG(ShareWomen) AS Mean_women,
       SUM(Total) AS Total_graduates
  FROM new_grads
 GROUP BY Major_category, Sample_category
HAVING Mean_women > 0.8;
```

Note that we used the same column name in the `HAVING` statement that we originally specified with the `AS` statement. The statement above will result in the following output:

Major_category|	Sample_category|	Mean_women|	Total_graduates
---|---|---|---|
Education|	Large	|0.923745479|	170862
Health|	Large|	0.896018988|	209394
Psychology & Social Work|	Medium|	0.81070414753552|	53552

Note that the results only include categories where `Mean_women` is greater than `.8`. That's because the `HAVING` statement filters out the other rows.





**Instructions:**

Remember to use the `new_grads` table.

- Find all of the major categories where the share of graduates with low-wage jobs is greater than `.1`.
  - Use the `SELECT` statement to select `Major_category` and `AVG(Low_wage_jobs)` / `AVG(Total)` as `Share_low_wage`
  - Use the `GROUP BY` statement to group the query by the `Major_category` column.
  - Use the `HAVING` statement to restrict the selection to rows where `Share_low_wage` is greater than `.1`.

```
SELECT Major_category,
       AVG(Low_wage_jobs) / AVG(Total) AS 'Share_low_wage'
  FROM new_grads
 GROUP BY Major_category
HAVING Share_low_wage > 0.1;
```

[7 rows x 2 columns]

Major_category|Share_low_wage
---|---
Arts|0.16833085991095678
Communications & Journalism|0.1263241815481876
Humanities & Liberal Arts|0.13208721344194835
Industrial Arts & Consumer Services|0.11571334076033978
Law & Public Policy|0.11568503743572278

## 9. Order of Execution

In the previous mission we learned that when executing a SQL query, the computer runs the clauses in this order:

1. `FROM`
2. `WHERE`
3. `SELECT`
4. `ORDER BY`
5. `LIMIT`

We now learned a couple more clauses: `GROUP BY` and `HAVING`. We can expand our mental model of the general structure of a query:

```
SELECT column(s)
  FROM some_table
 WHERE some_condition
 GROUP BY column(s)
HAVING some_condition
 ORDER BY column(s)
 LIMIT some_limit;

```

And the order by which SQL runs this is:

1. `FROM`
2. `WHERE`
3. `GROUP BY`
4. `HAVING`
5. `SELECT`
6. `ORDER BY`
7. `LIMIT`

Note that `ORDER BY` is executed after `GROUP BY`. One of the main goals of ordering results is having functional output, so it's not unreasonable to think that it should be one of the last clauses to run.



## 10. Rounding Results With the ROUND() Function

On a previous screen, the percentages in our results were very long and hard to read (e.g., `0.16833085991095678`). We can use the SQL `ROUND` function in our query to round them. Here's an example of what this looks like:
```
SELECT Major_category,
       ROUND(ShareWomen, 2) AS rounded_share_women 
  FROM new_grads;
```

The query will round the `ShareWomen` column to two decimal places. Here's a truncated view of the results:

Major_category|	rounded_share_women
---|---
Engineering|	0.12
Engineering	|0.1

By passing different values into the `ROUND` function, such as `ROUND(ShareWomen, 3)`, we can round to different decimal places.

**Instructions:**

Write a query with the following features:
- Displays, in this order:
  - ShareWomen rounded to 4 decimal places with the alias Rounded_women
  - Major_category
- Displays only the first ten results resulting from the above

```
SELECT ROUND(ShareWomen, 4) AS Rounded_women, Major_category
  FROM new_grads
 LIMIT 10;
```

[10 rows x 2 columns]

Rounded_women|Major_category
---|---
0.1206|Engineering
0.1019|Engineering
0.153|Engineering
0.1073|Engineering
0.3416|Engineering
0.145|Engineering
0.5357|Business
0.4414|Physical Sciences
0.1398|Engineering
0.4378|Engineering

## 11. Nesting functions

On a previous screen, we ran the following query:
```
SELECT Major_category, Sample_category,
       AVG(ShareWomen) AS Mean_women,
       SUM(Total) AS Total_graduates
  FROM new_grads
 GROUP BY Major_category, Sample_category
HAVING Mean_women > 0.8;
```
This query returned very long fractional values for `Mean_women`. We can update our query with the `ROUND` function to round the results to three decimal places:

```
SELECT Major_category, Sample_category,
       ROUND(AVG(ShareWomen), 3) AS Mean_women,
       SUM(Total) AS Total_graduates
  FROM new_grads
 GROUP BY Major_category, Sample_category
HAVING Mean_women > 0.8;
```

This will return the following result:

Major_category|	Sample_category|	Mean_women|	Total_graduates
---|---|---|---
Education|	Large|	0.924|	170862
Health|	Large|	0.896|	209394
Psychology & Social Work	|Medium|	0.811|	53552


**Instructions:**

Write a query that:
- Aggregates by the major's category
- Displays in order:
  - The major's category
  - The ratio `AVG(College_jobs)` / `AVG(Total)` rounded to three decimal places with the alias `Share_degree_jobs`
- Only keeps the rows where `Share_degree_jobs` is less than `.3`

```
SELECT Major_category,
       ROUND(AVG(College_jobs) / AVG(Total), 3) AS 'Share_degree_jobs'
  FROM new_grads
 GROUP BY Major_category
HAVING Share_degree_jobs < .3;
```

[8 rows x 2 columns]

Major_category|Share_degree_jobs
---|---
Agriculture & Natural Resources|0.248
Arts0.265Business|0.114
Communications & Journalism|0.22
Humanities & Liberal Arts|0.27
Industrial Arts & Consumer Services|0.249

## 12. Casting

On the last few screens, we used SQL arithmetic to divide float (what we referred to as decimal numbers in the previous mission) columns. This resulted in float values that we could round using the `ROUND()` function.

If we try to divide two integer columns (`Women` and `Total`, for instance), SQLite (and most other SQL dialects) will round down and return integer values:

```
SELECT Women, Total,
       Women/Total SW
  FROM new_grads
 LIMIT 12;
```

This query returns:


|Women|	Total	|SW|
|---|---|---|
|282|	2339|	0|
77	|756|	0
131	|856|	0
135	|1258|	0
11021	|32260|	0
373	|2573|	0
960|	3777|	0
1667	|1792|	0
2105	|91227|	0
6548	|81527|	0
8284	|41542|	0
16016	|15058|	1

Notice how everytime `Women` is smaller than `Total` the result `0` and when Women is larger, the result is `1`. The ratio was rounded.

To get float value, we can use the `CAST()` function to the transform the columns into `Float` type:

```
SELECT CAST(Women AS Float) / CAST(Total AS Float) AS women_ratio
  FROM new_grads 
 LIMIT 5;
```
This returns the results as float values:

|women_ratio| |
|---|---|
|0.12056434373663959|
0.10185185185185185
0.1530373831775701
0.10731319554848967
0.3416305021698698


**Instructions:**

Write a query that
- Divides the sum of the `Women` column by the sum of the `Total` column, aliased as `SW` and as float values.
- Group the results by `Major_category` and order by `SW`.
- Only contains the `Major_category` and `SW` columns, in that order.


```
SELECT Major_category, 
       CAST(SUM(Women) AS FLOAT) / CAST(SUM(Total) AS FLOAT) AS 'SW'
  FROM new_grads
 GROUP BY Major_category
 ORDER BY SW;
 ```

 [16 rows x 2 columns]

Major_category|SW
---|---
Law & Public Policy|0.030585069260274586
Business|0.08474280852841269
Industrial Arts & Consumer Services|0.16024926890405236
Computers & Mathematics|0.20935560252568494
Engineering|0.2195958577559186

## 13. Next Steps

In this mission, we covered the CASE, GROUP BY and HAVING clauses. We can use these to quickly calculate powerful summary statistics in SQL and program if/then logic. In the next few missions, we'll learn more about working with SQL tables, including how to insert and modify data.